gsy_framework/sim_results/electric_blue/volume_timeseries.py

```diff
@@ -0,0 +1,167 @@
+        """Save asset volume time series in the DB."""
+        # TODO: should be implemented.
+        raise NotImplementedError()
+
+    def _add_total_energy_bought(
+            self, asset_stats: ForwardDeviceStats, market_type: AvailableMarketTypes):
+        """Add statistics for the total amount of bought energy."""
+        energy_kWh = asset_stats.total_energy_bought
+        if energy_kWh <= 0:
+            return
+        profile = self._trade_profile_generator.generate_trade_profile(
+            energy_kWh, asset_stats.time_slot, market_type)
+        for time_slot, value in profile.items():
+            self._add_to_volume_time_series(time_slot, value, market_type, "bought_kWh")
+
+    def _add_total_energy_sold(
+            self, asset_stats: ForwardDeviceStats, market_type: AvailableMarketTypes):
+        """Add statistics for the total amount of sold energy."""
+        energy_kWh = asset_stats.total_energy_sold
+        if energy_kWh <= 0:
+            return
+        profile = self._trade_profile_generator.generate_trade_profile(
+            energy_kWh, asset_stats.time_slot, market_type
+        )
+        for time_slot, value in profile.items():
+            self._add_to_volume_time_series(time_slot, value, market_type, "sold_kWh")
+
+    def _add_to_volume_time_series(
+            self, time_slot: DateTime, value: float,
+            market_type: AvailableMarketTypes, attribute_name: str):
+        """Add time slot value to the correct time slot to asset volume time series."""
+        time_slot = self._adapt_time_slot(time_slot)
+        volume_time_series = self._get_asset_volume_time_series(
+            year=time_slot.start_of("year"))
+        volume_time_series[time_slot][market_type.name][attribute_name] += value
+
+    def _adapt_time_slot(self, time_slot: DateTime) -> DateTime:
+        """Find the containing time slot of the smaller time slot."""
+        if self.resolution == AggregationResolution.RES_15_MINUTES:
+            return time_slot.set(minute=(time_slot.minute // 15) * 15)
+        return time_slot.start_of(START_OF[self.resolution])
+
+    def _get_asset_volume_time_series(self, year: DateTime):
+        """Return asset volume time series for the required year.
+        If not found in the buffer, it tries fetching it from DB.
+        If not found in the DB, it will generate a new one for the whole year."""
+        if year in self._asset_volume_time_series_buffer:
+            time_series = self._asset_volume_time_series_buffer[year]
+        else:
+            time_series = self._fetch_asset_volume_time_series_from_db(year)
+            if time_series is None:
+                time_series = {
+                    ts: self._get_time_series_template(value)
+                    for ts, value in self._generate_SSP_time_series(year)}
+            self._asset_volume_time_series_buffer[year] = time_series
+        return time_series
+
+    def _fetch_asset_volume_time_series_from_db(
```

Comment:
Could you elaborate on this method, how do you see the implementation for reading and writing to db? Since gsy-web is responsible for interacting with db, we were using `ResultsHandler` and it's [update](https://github.com/gridsingularity/gsy-framework/blob/e10b82e797ab44bfa612d4a81b26bc43dd315dfa/gsy_framework/sim_results/all_results.py#L71) method to transfer data. We provide it with the core stats, and we could provide it with the pre-fetched data, but do you expect to have more interaction with db, like query it from this method?

Reply:
This method needs to fetch data from DB. So, yes! You need to run queries from this class. The class also needs to save the updated results by another query. Pre-fetching can't be a solution because we don't know what to fetch at the beginning.